### PR TITLE
Refactoring of ScanHub Workflows

### LINKED ISSUE #90: Refactoring of Jobs/Workflows
With the input from the A4IM Hackathon in Freiburg the Job was refactored.
A job now contains a list of tasks. A task can either describe a data acquisition step (device-task) or a processing step (processing-step). The job database model has been modified accordingly and a new table for tasks has been implemented:

```
class Task(Base):
    """Task ORM model."""

    __tablename__ = "task"
    __table_args__ = {"extend_existing": True}

    # Use uuid here
    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    description: Mapped[str] = mapped_column(nullable=False)
    type: Mapped[TaskType] = mapped_column(type_=JSON, nullable=False)

    # Arguments and parameters
    # Example: "args": {"arg1": "x", "arg2": "y"}
    args: Mapped[dict[str, str]] = mapped_column(type_=JSON, nullable=True)
    # acquisition_limits: Mapped[AcquisitionLimits] = mapped_column(type_=JSON, nullable=False)
    # sequence_parameters: Mapped[SequenceParameters] = mapped_column(type_=JSON, nullable=False)
    # device_parameters: Mapped[DeviceParameters] = mapped_column(type_=JSON, nullable=False)


    # Input and output artifacts with export destination
    # Example: {"input": [{"name": "env_HOLOSCAN_INPUT_PATH", "value": "{{ context.input.dicom }}"}]}
    # https://github.com/Project-MONAI/monai-deploy/blob/main/deploy/monai-deploy-express/sample-workflows/hello-world.json
    artifacts: Mapped[dict[str, list[dict[str, str]]]] = mapped_column(type_=JSON, nullable=True) # implemention of input and output artifacts for artifact types as replacement for string

    # List of task destinations, which are for example used to create the chain of topics in the kafka message broker, i.e., target topics
    task_destinations: Mapped[list[dict[str, str]]] = mapped_column(type_=JSON, nullable=True)

    # Relations and references
    job_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("job.id"))

    status: Mapped[dict[TaskStatus, str]] = mapped_column(type_=JSON, nullable=False)

    # Fields
    datetime_created: Mapped[datetime.datetime] = mapped_column(
        server_default=func.now()  # pylint: disable=not-callable
    )
```
 
### ToDo's

- [x] Task database table
- [x] Task pydantic model
- [x] Task data acess layer
- [x] Task endpoints
- [x] Update Job
- [x] Remove obsolete models
- [x] Consequently use models defined in shared_lib instead of exam-manager models

### Notes

- Output/input artifacts are in stored in a shared datalake, parent folder corresponds to job-id, subfolder can be defined in artifact
- Inputs and outputs are described by artifacts, i.e. they contain the file paths
- Arguments are specific for devices/acquisitions, it might be hard to fully describe them in advance (by using a pedantic model). For now it has been decided to pass any parameters as a dictionary contained in `args`
- Some of the job parameters has been moved to the patient database (weight, height, gender). When performing an acquisition, only the parameters relevant for the SAR calculation should be obtainable, i.e. patient data must not be accessible by the workflow-manager and acquisition-control. Could be solved by patient endpoint for SAR endpoint

### LINKED ISSUE #91: Refactoring of Workflow-Engine/Acquisition-Control
The job has been refactored and can now contain device-tasks (acquisitions) and processing steps (reconstructions). The workflow-engine is the central connection point for the UI and receives a list of tasks (a job). It parses all tasks and distributes them accordingly (trigger kafka-topic or acquisition-control). To distinguish different tasks, each task has a type.

ToDo's:

- [ ] Parse task list from job
- [ ] Update acquisition control with new task syntax (i.e. replace scanjob, acquisition parameter, acquisition limits)
- [ ] Workflow endpoint for job execution
- [ ] Workflow endpoints which returns all informations
- [ ] Remove obsolete workflow-service

### LINKED ISSUE #93: Implementation of Exam/Workflow/Task Templates
Implementation of Template Tables/Models:

- [x] Exam template
- [x] Job template
- [x] Task template

Separate tables for template and concrete jobs.

### LINKED ISSUE #94: Publishing of ScanHub Base Image
Regularly publish the Docker base image of ScanHub to GitHub Packages or to Docker hub.
Information on how to setup the GitHub action can be found here:
https://docs.github.com/en/actions/publishing-packages/publishing-docker-images 

Preferred solution would be GitHub Packages, to have the base image with the code.
All Dockerfiles which use the base image should be updated to use the latest image version from GitHub Package.

### ToDo's:

- [ ] Implementation and testing of GitHub action to publish base image
- [ ] Use published base image in Dockerfiles which are build on the base image

### PR DESCRIPTION
**Pull request by schote**
This pull-request refactors the overall job/workflow/acquisition pipeline and includes the following sub-tasks:

- [x] Refactoring of job description and implementation of tasks, 
closes #90 
- [x] Implementation of templates for exams, jobs and tasks, 
closes #93 
- [x] Remove obsolete workflow and record tables from database

### CODE REVIEW

**Comment by schote** [comment]
Response body of an exam including jobs and tasks:
```
[
  {
    "patient_id": 0,
    "name": "string",
    "country": "string",
    "site": "string",
    "address": "string",
    "creator": "string",
    "status": "string",
    "id": "1523d989-aa29-43e0-9760-807ecd4e68bb",
    "datetime_created": "2024-02-24T09:09:16.560759",
    "datetime_updated": null,
    "jobs": [
      {
        "comment": "string",
        "exam_id": "1523d989-aa29-43e0-9760-807ecd4e68bb",
        "is_finished": true,
        "id": "d6de6bea-0948-413f-98c0-cb6c7fb1d1e0",
        "tasks": [
          {
            "job_id": "d6de6bea-0948-413f-98c0-cb6c7fb1d1e0",
            "description": "string",
            "type": "PROCESSING_TASK",
            "args": {
              "additionalProp1": "string",
              "additionalProp2": "string",
              "additionalProp3": "string"
            },
            "artifacts": {},
            "task_destinations": [
              {
                "additionalProp1": "string",
                "additionalProp2": "string",
                "additionalProp3": "string"
              }
            ],
            "status": {},
            "id": "8b08f7ab-a7ed-4c15-9303-286ec1f12c4d",
            "datetime_created": "2024-02-24T10:10:56.520844"
          },
          {
            "job_id": "d6de6bea-0948-413f-98c0-cb6c7fb1d1e0",
            "description": "string",
            "type": "PROCESSING_TASK",
            "args": {
              "additionalProp1": "string",
              "additionalProp2": "string",
              "additionalProp3": "string"
            },
            "artifacts": {},
            "task_destinations": [
              {
                "additionalProp1": "string",
                "additionalProp2": "string",
                "additionalProp3": "string"
              }
            ],
            "status": {},
            "id": "31b0d7b3-4868-4122-a554-b3d66f22ea55",
            "datetime_created": "2024-02-24T10:12:46.697199"
          },
          {
            "job_id": "d6de6bea-0948-413f-98c0-cb6c7fb1d1e0",
            "description": "string",
            "type": "PROCESSING_TASK",
            "args": {
              "additionalProp1": "string",
              "additionalProp2": "string",
              "additionalProp3": "string"
            },
            "artifacts": {},
            "task_destinations": [
              {
                "additionalProp1": "string",
                "additionalProp2": "string",
                "additionalProp3": "string"
              }
            ],
            "status": {
              "PENDING": "test"
            },
            "id": "1a1050e1-fecd-4eda-b334-2a21ec07a0b7",
            "datetime_created": "2024-02-24T10:18:33.413817"
          },
          {
            "job_id": "d6de6bea-0948-413f-98c0-cb6c7fb1d1e0",
            "description": "string",
            "type": "PROCESSING_TASK",
            "args": {
              "additionalProp1": "string",
              "additionalProp2": "string",
              "additionalProp3": "string"
            },
            "artifacts": {
              "input": [],
              "output": []
            },
            "task_destinations": [
              {
                "additionalProp1": "string",
                "additionalProp2": "string",
                "additionalProp3": "string"
              }
            ],
            "status": {
              "PENDING": "test"
            },
            "id": "9878f812-d0cd-46c9-991a-cc74d24b248c",
            "datetime_created": "2024-02-24T10:19:32.294834"
          },
          {
            "job_id": "d6de6bea-0948-413f-98c0-cb6c7fb1d1e0",
            "description": "string",
            "type": "PROCESSING_TASK",
            "args": {
              "additionalProp1": "string",
              "additionalProp2": "string",
              "additionalProp3": "string"
            },
            "artifacts": {
              "input": [
                {
                  "path": "test",
                  "name": "coil-map"
                },
                {
                  "path": "test2",
                  "name": "k-space"
                }
              ],
              "output": []
            },
            "task_destinations": [
              {
                "additionalProp1": "string",
                "additionalProp2": "string",
                "additionalProp3": "string"
              }
            ],
            "status": {
              "PENDING": "test"
            },
            "id": "807dd0ee-4c47-4ac7-8e17-d344b49e4c1c",
            "datetime_created": "2024-02-24T10:23:17.721004"
          }
        ],
        "datetime_created": "2024-02-24T09:09:49.197812",
        "datetime_updated": null
      }
    ]
  }
]
```

**Comment by chdinh** [comment]
Great job David thx 👍

File: infrastructure/nginx_config.conf [hunk 1]
```diff
@@ -2,9 +2,9 @@ server {
   listen 8080;
   root /etc/nginx/html;
 
-  location /api/v1/workflow {
-    proxy_pass http://workflow-manager:8000/api/v1/workflow;
-  }
+  # location /api/v1/workflow {
+  #   proxy_pass http://workflow-manager:8000/api/v1/workflow;
+  # }
```
Comment (chdinh) [inline infrastructure/nginx_config.conf:7]: is this commented out on purpose?

File: infrastructure/nginx_config.conf [hunk 2]
```diff
@@ -26,7 +26,7 @@ server {
     proxy_pass http://sequence-manager:8000/api/v1/mri/sequences;
   }
 
-  location /api/v1/mri/acquisitioncontrol {
-    proxy_pass http://acquisition-control:8000;
-  }
+  # location /api/v1/mri/acquisitioncontrol {
+  #   proxy_pass http://acquisition-control:8000;
+  # }
```
Comment (chdinh) [inline infrastructure/nginx_config.conf:31]: is this commented out on purpose?

File: services/exam-manager/app/main.py [hunk 3]
```diff
@@ -54,11 +54,11 @@ async def startup():
...
-    if "workflow" not in tables:
-        raise HTTPException(
-            status_code=500,
-            detail="SQL-DB: Workflow table is required but does not exist.",
-        )
+    # if "workflow" not in tables:
+    #     raise HTTPException(
+    #         status_code=500,
+    #         detail="SQL-DB: Workflow table is required but does not exist.",
+    #     )
```
Comment (chdinh) [inline services/exam-manager/app/main.py:61]: if check guard is not enabled on purpose?

**Review by chdinh — CHANGES_REQUESTED** [review]

File: infrastructure/nginx_config.conf (hunk 1 above)
Comment (schote) [reply infrastructure/nginx_config.conf:7]: Thank you! Forgot abou this, it was just for development purposes

File: infrastructure/nginx_config.conf (hunk 2 above)
Comment (schote) [reply infrastructure/nginx_config.conf:31]: See above

File: services/exam-manager/app/main.py (hunk 3 above)
Comment (schote) [reply services/exam-manager/app/main.py]: This is inside the startup routine of the exam-manager and was used to check if the workflow table (created by the workflow-service) has been created successfully. As this table is obsolete now and has been removed (we only have the workflow table created by the exam-manager, which was the Job table previously) this check is not feasible anymore. Instead we check the existence of the workflow-table in the health check endpoint.

**Review by chdinh — APPROVED** [review]

**Pull request merged**